### PR TITLE
Add USTONKS-0921 price identifier

### PR DESCRIPTION
**Pull request by rossgalloway**
New UMIP for uSTONKS-0921 with a new basket of tickers.

Comments and questions:
- The linked spreadsheet should start working once the feed from googlefinance updates. That may be on 6/16/2021.
- Should this PID (and the contract) be 6 decimals to match USDC? 
- The price feed implementation link on line 58 is for uniswap. We will be pooling liquidity on sushiswap. Is there a different implementation for sushiswap?

### CODE REVIEW

File: UMIPs/UMIP add uStonks-0921.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+- Do these sources allow for querying up to 74 hours of historical data? Yes
+- How often is the provided price updated? The provided endpoint queries for prices on a daily basis. This is the level of granularity needed to determine the expiry price.
+- Is an API key required to query these sources? No.
+- Is there a cost associated with usage? No.
+- If there is a free tier available, how many queries does it allow for? NA
+- What would be the cost of sending 15,000 queries? $0.
+
+## PRICE FEED IMPLEMENTATION
+For the creation of the uSTONKS token, it is desired that the DVM return either the final closing index value of uSTONKS, or a 2-hour TWAP on the market price of uSTONKS_JUN21. The type of price that the DVM will return is dependent on the timestamp the price request is made at. This timestamp is the expiry timestamp of the contract that is intended to use this price identifier, so the TWAP calculation is used pre-expiry and the closing index value of uSTONKS calculation is used at expiry.  This is very similar to the uGAS token and this design is outlined in [UMIP 22](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-22.md).
```
Comment (Reinis-FRP) [inline UMIPs/UMIP add uStonks-0921.md:54]: uSTONKS_JUN21 -> uSTONKS_0921

File: UMIPs/UMIP add uStonks-0921.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+## IMPLEMENTATION
+Voters should determine which pricing implementation to use depending on when the price request was submitted.
+
+**At Expiry**
+
+If the price request's UTC timestamp is at 1633046400 (Oct. 1, 2021 00:00:00 UTC), a price request for uSTONKS for a given timestamp should be determined by performing the `At Expiry` process.
+
+To create a baseline for the uSTONKS index, a base price for each stock was chosen. These prices are the closing prices of each on 06/15/2021 and can be used by voters as reference values for their calculations. Voters are highly encouraged to verify these values themselves. This basket is static - it will not change over the life of the contract.
+
+| Ticker | Stock Price 04/14/2021 |
```
Comment (Reinis-FRP) [inline UMIPs/UMIP add uStonks-0921.md:86]:
```
 | Ticker | Stock Price 06/15/2021 |
```

File: UMIPs/UMIP add uStonks-0921.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+2. Divide the September 30 price by the base price and multiply by the index value base (10) to get today’s index value. 
+3. Perform this function for each component in the index and sum all of the results together.
+4. Round the result of step three to six decimal places
+
+An example of this calculation is shown in [this](https://docs.google.com/spreadsheets/d/1IbvIZUGQG2TW9YLrV7ykPr3P2F4PhvgvOU-4o2RujZc/edit?usp=sharing) Google Sheet.
+
+It is possible that, over the life of this price identifier, certain situations could arise that would affect the validity of the index calculation. Guidance is provided below for certain situations, but voters are free to come to an alternative consensus if a flaw in the methodology is identified or an unforeseen event occurs. 
+
+1. In the case of a stock split, the base price of the stock should be divided by the number of shares that the stock is being split into. As an example - if TSLA issued a 4:1 stock split, the base price of TSLA, 732.23, would be divided by 4 and the new base index price would be 183.06.
```
Comment (Reinis-FRP) [inline UMIPs/UMIP add uStonks-0921.md:111]: TSLA is not a part of uSTONKS_0921 basket

File: UMIPs/UMIP add uStonks-0921.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+2. In the case of a stock consolidation, the process is the same as for a stock split, except the base price should be multiplied by the number of shares being consolidated. 
+
+**Before Expiry**
+
+If the price request's UTC timestamp is less than 1633046400 (Oct 1, 2021 00:00:00 UTC), voters will need to calculate a 2-hour TWAP for the uSTONKS_0921 token’s price in USDC. The following process should be used to calculate the TWAP.
+1. The end TWAP timestamp equals the price request timestamp.
+2. The start TWAP timestamp is defined by the end TWAP timestamp minus the TWAP period (2 hours in this case).
+3. A single Sushiswap price is defined for each timestamp as the price that the uSTONKS_JUN21 / USDC pool returns in the latest block where the block timestamp is <= the price request timestamp. This is the price of 1 uSTONKS_JUN21 token in USDC.
```
Comment (Reinis-FRP) [inline UMIPs/UMIP add uStonks-0921.md:119]: uSTONKS_JUN21 -> uSTONKS_0921

File: UMIPs/UMIP add uStonks-0921.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
...tunity for sponsors to mint tokens with less collateral than what they can sell them from in the market. It is important to note that this is an edge case scenario either driven by an irrational change in market expectations or it can be driven by a “fat finger” mistake which is a vulnerability to any market. Even in this edge case we believe the design of the token and the parameters chosen should mitigate risks. The current Expiring Multi Party (EMP) contract requires sponsors to mint tokens with enough collateral to meet the Global Collateral Ratio (GCR) which has stood well above 200% for other contracts. Therefore, assuming the GCR is similar for uSTONKS_APR21, the market would need to first rally at least 100% before potentially being exposed. If the sponsor wishes to withdraw collateral below the GCR they would request a “slow withdrawal” which would subject him to a 2 hour “liveness period” where anybody can liquidate the position if it fell below the required collateral ratio. The combination of the GCR and 2 hour “liveness period” allows the 2 hour TWAP to “catch up” to the market price and would protect from this scenario and deter sponsors from attempting to exploit it.
```
Comment (Reinis-FRP) [inline UMIPs/UMIP add uStonks-0921.md:128]: uSTONKS_APR21 -> uSTONKS_0921

File: UMIPs/UMIP add uStonks-0921.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+| Created    | June 11, 2021   
+| Discourse Link | https://discourse.umaproject.org/t/add-ustonks-0921-price-identifer/1195 |  
+
+## SUMMARY
+
+The DVM should support price requests for an updated basket of stocks that represent the top ten tickers with the most comment volume on r/wallstreetbets. These ten stocks will be equally weighted to create an index named uSTONKS. 
+
+## MOTIVATION
+
+Update the existing uStonks Price identifier (UMIP-79) used by the DMV to support a new basket of stocks based on the current sentiment on the r/wallstreetbets subreddit.
```
Comment (Reinis-FRP) [inline UMIPs/UMIP add uStonks-0921.md:16]: If it is expected to update this on a quarterly basis, maybe it is worth considering creating a generalized uStonks price identifier where relevant parameters can be passed as ancillary data:
- expiration timestamp - to be used by FPL price identifier switcher
- index components (markets & symbols) - for expiration
- AMM pool address - for pre-expiration

Comment (rossgalloway) [reply UMIPs/UMIP add uStonks-0921.md:16]: Yes, agreed that this is something that we should do for both uSTONKS and uGAS. After this round I will create a new one that does this.

File: UMIPs/UMIP add uStonks-0921.md (hunk 1 above)
Comment (rossgalloway) [reply UMIPs/UMIP add uStonks-0921.md]: Should have now corrected all the names now.

File: UMIPs/UMIP add uStonks-0921.md (hunk 2 above)
Comment (rossgalloway) [reply UMIPs/UMIP add uStonks-0921.md]: fixed

File: UMIPs/UMIP add uStonks-0921.md (hunk 3 above)
Comment (rossgalloway) [reply UMIPs/UMIP add uStonks-0921.md]: changed to GME

**Review by therealhexi — APPROVED** [review]
Looks good to community reviewer

File: UMIPs/UMIP add uStonks-0921.md [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+If the price request's UTC timestamp is less than 1633046400 (Oct 1, 2021 00:00:00 UTC), voters will need to calculate a 2-hour TWAP for the uSTONKS_0921 token’s price in USDC. The following process should be used to calculate the TWAP.
+1. The end TWAP timestamp equals the price request timestamp.
+2. The start TWAP timestamp is defined by the end TWAP timestamp minus the TWAP period (2 hours in this case).
+3. A single Sushiswap price is defined for each timestamp as the price that the uSTONKS_0921 / USDC pool returns in the latest block where the block timestamp is <= the price request timestamp. This is the price of 1 uSTONKS_0921 token in USDC. 
+4. The TWAP is an average of the prices for each timestamp between the start and end timestamps. Each price in this average will get an equal weight.
+5. This results should be rounded to 6 decimals.
+
+For both implementations, voters should determine whether the returned price differs from broad market consensus. This is meant to provide flexibility in any unforeseen circumstances as voters are responsible for defining broad market consensus.
+
+## Security Considerations
```
Comment (smb2796) [inline UMIPs/UMIP add uStonks-0921.md:125]: To keep this simpler and reduce redundancy, I'd suggest just referring to previous uSTONKS umips for the security considerations and rationale sections, as these look to be the exact same content.

Comment (mhairimca) [reply UMIPs/UMIP add uStonks-0921.md:125]: Wouldnt it be better just to have the copy and paste rather than create unnecessary dependencies

Comment (smb2796) [reply UMIPs/UMIP add uStonks-0921.md:125]: I don't think so, but guess it doesn't matter too much!

File: UMIPs/UMIP add uStonks-0921.md (hunk 6 above)
Comment (smb2796) [reply UMIPs/UMIP add uStonks-0921.md:16]: The issue with doing this now is that uSTONKS and uGAS both use the EMP, which does not support custom ancillary data at this time.

So what could be done is - separating out the self-referential and rolling median prices and using a FPL to switch these at expiry. But this doesn't really save UMIPs, as the self-referential pid would still need to be re-proposed without being able to use custom ancillary data.

**Review by smb2796 — APPROVED** [review]
Looks good. Merging in and moving to Last Call

**Review by CarelessDoge — APPROVED** [review]

**Pull request merged**